File: pkg/spec/test/host-parity.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { test } from "node:test";

import { builtinCandidates } from "../src/pick.ts";
import { resetWarnings } from "../src/verbs.ts";
import type { BuiltinVerb } from "../src/action-tree.ts";
import { fakeElement, withFakeDocument, type FakeElementSpec } from "./web-dom-harness.ts";
import { __testing__ } from "../src/web-runtime.ts";

const { host } = __testing__;

const goldenPath = fileURLToPath(new URL("./fixtures/host-parity-golden.json", import.meta.url));
const golden: Record<string, string[]> = JSON.parse(readFileSync(goldenPath, "utf8"));

const VERBS: BuiltinVerb[] = [
  "taps",
  "doubleTaps",
  "longPresses",
  "typing",
  "scrolls",
  "swipes",
  "pressKeys",
  "waitOnce",
];

// SCREEN is the canonical screen both hosts are driven over: one row per fact
// combination that any verb distinguishes. The native host builds the same rows,
// in the same order, as a hierarchy tree in
// internal/verifier/host_parity_test.go.
const SCREEN: (FakeElementSpec & { name: string })[] = [
  { name: "root", tag: "html", x: 0, y: 0, width: 400, height: 800, overflows: true },
  { name: "save", tag: "button", x: 0, y: 0, width: 200, height: 60, clickable: true },
  {
    name: "cancel",
    tag: "button",
    x: 200,
    y: 0,
    width: 200,
    height: 60,
    clickable: true,
    disabled: true,
  },
  { name: "amount", tag: "input", x: 0, y: 100, width: 400, height: 60, editable: true },
  { name: "list", tag: "div", x: 0, y: 200, width: 400, height: 400, overflows: true },
  { name: "row", tag: "div", x: 0, y: 600, width: 400, height: 80 },
  { name: "collapsed", tag: "button", x: 0, y: 0, width: 0, height: 0, clickable: true },
];

// The web host and the native host used to route verbs themselves and had
// drifted: web sent `swipes` to scrollable containers only, so a swipe on a list
// row was reachable on Android and unreachable on web for the same spec.
//
// Per-verb eligibility now has ONE definition (src/targets.ts); a host reports
// facts and never filters. This test is what notices if a second definition
// grows back on either side. internal/verifier/host_parity_test.go asserts the
// SAME golden from the native host, so a match on both sides proves the two
// hosts agree without either invoking the other.
test("web host targets match the cross-host golden, verb for verb", () => {
  const elements = SCREEN.map(fakeElement);
  withFakeDocument(elements, () => {
    for (const verb of VERBS) {
      resetWarnings();
      const candidates = builtinCandidates(verb, host);
      assert.notEqual(candidates.length, 0, `${verb} enumerated nothing at all`);
      const named = candidates
        .filter((candidate) => candidate.targetIndex >= 0)
        .map((candidate) => SCREEN[candidate.targetIndex]!.name);
      assert.deepEqual(named, golden[verb], `web host targets for ${verb}`);
    }
  });
});

File: internal/verifier/host_parity_test.go
package verifier

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"slices"
	"testing"
)

// hostParityScreen is the canonical screen both hosts are driven over: one row
// per fact combination that any verb distinguishes. Each host builds it in its
// own model (this file as a hierarchy tree, pkg/spec/test/host-parity.test.ts as
// a DOM), enumerates every verb, and asserts the SAME committed golden.
var hostParityScreen = []struct {
	name           string
	clickable      bool
	enabled        bool
	editable       bool
	scrollable     bool
	positiveBounds bool
}{
	{name: "root", enabled: true, scrollable: true, positiveBounds: true},
	{name: "save", clickable: true, enabled: true, positiveBounds: true},
	{name: "cancel", clickable: true, positiveBounds: true},
	{name: "amount", enabled: true, editable: true, positiveBounds: true},
	{name: "list", enabled: true, scrollable: true, positiveBounds: true},
	{name: "row", enabled: true, positiveBounds: true},
	{name: "collapsed", clickable: true, enabled: true},
}

// hostParityTreeJSON is hostParityScreen as the native host sees it. Tree order
// is pre-order, so it matches hostParityScreen index for index.
const hostParityTreeJSON = `{
  "attributes": {"resource-id": "root", "scrollable": "true", "bounds": "[0,0,400,800]"},
  "enabled": true,
  "children": [
    {"attributes": {"resource-id": "save", "bounds": "[0,0,200,60]"}, "clickable": true, "enabled": true, "children": []},
    {"attributes": {"resource-id": "cancel", "bounds": "[200,0,400,60]"}, "clickable": true, "enabled": false, "children": []},
    {"attributes": {"resource-id": "amount", "bounds": "[0,100,400,160]"}, "editable": true, "enabled": true, "children": []},
    {"attributes": {"resource-id": "list", "scrollable": "true", "bounds": "[0,200,400,600]"}, "enabled": true, "children": []},
    {"attributes": {"resource-id": "row", "bounds": "[0,600,400,680]"}, "enabled": true, "children": []},
    {"attributes": {"resource-id": "collapsed", "bounds": "[0,0,0,0]"}, "clickable": true, "enabled": true, "children": []}
  ]
}`

// TestHostsAgreeOnTargetEligibility is the guard on the claim that one
// specification induces one action space on every platform. The native host and
// the web host used to route verbs themselves and had drifted: web sent
// `swipes` to scrollable containers only, so a swipe on a list row was
// reachable on Android and unreachable on web for the same spec.
//
// Per-verb eligibility now has ONE definition (pkg/spec/src/targets.ts); a host
// reports facts and never filters. This test is what notices if a second
// definition grows back on either side: both hosts enumerate the same canonical
// screen and must name the same targets, verb for verb, in the same order.
// pkg/spec/test/host-parity.test.ts asserts the SAME golden from the web host,
// so a match on both sides proves the two hosts agree without either invoking
// the other.
func TestHostsAgreeOnTargetEligibility(t *testing.T) {
	golden := loadHostParityGolden(t)
	for _, verb := range policyVerbs {
		t.Run(verb, func(t *testing.T) {
			want, ok := golden[verb]
			if !ok {
				t.Fatalf("golden has no entry for %s", verb)
			}
			verifier := newVerifier(t, WithSeed(0x5eed))
			loadActionSpec(t, verifier, fmt.Sprintf(
				"import { %s } from \"@sanderling/spec\";\nglobalThis.actions = %s;", verb, verb))
			pushTree(t, verifier, hostParityTreeJSON)

			entries, err := verifier.enumerateBuiltin(verb)
			if err != nil {
				t.Fatalf("enumerate %s: %v", verb, err)
			}
			if len(entries) == 0 {
				t.Fatalf("%s enumerated nothing at all", verb)
			}
			got := []string{}
			for _, entry := range entries {
				if entry.targetIndex < 0 {
					continue
				}
				if entry.targetIndex >= len(hostParityScreen) {
					t.Fatalf("%s produced target index %d, off the %d-row screen",
						verb, entry.targetIndex, len(hostParityScreen))
				}
				got = append(got, hostParityScreen[entry.targetIndex].name)
			}
			if !slices.Equal(got, want) {
				t.Errorf("native host targets for %s\n got=%v\nwant=%v", verb, got, want)
			}
		})
	}
}

func loadHostParityGolden(t *testing.T) map[string][]string {
	t.Helper()
	path, err := filepath.Abs("../../pkg/spec/test/fixtures/host-parity-golden.json")
	if err != nil {
		t.Fatal(err)
	}
	body, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read golden: %v", err)
	}
	golden := map[string][]string{}
	if err := json.Unmarshal(body, &golden); err != nil {
		t.Fatalf("decode golden: %v", err)
	}
	return golden
}
